### ISSUE TITLE
Allow Mobile platforms to get external app media playback state

### ISSUE THREAD

**Issue by ztc0611** [godotengine/godot-proposals#8582]
### Describe the project you are working on

A mobile app/game.

### Describe the problem or limitation you are having in your project

Godot games don't have the ability to react to the user's current audio besides to override it, or mix with it.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

On iOS, [there is a way](https://developer.apple.com/documentation/avfaudio/avaudiosession/1616600-secondaryaudioshouldbesilencedhi) to get if another app, such as Spotify, is playing audio. This [appears to be the case](https://developer.android.com/reference/android/media/session/MediaSessionManager) on Android as well, but I'm still investigating the specifics. This would allow the developer to do something such as muting the in-game soundtrack but keeping sound effects enabled if the user is already playing music.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This will add `NOTIFICATION_OS_EXTERNAL_MEDIA_BEGIN` and `NOTIFICATION_OS_EXTERNAL_MEDIA_END`, alongside `bool OS.is_media_playing()` in order to get the state of media playback on the device.

I already have these three functions operational on iOS already. This functionality is not available on macOS, so I doubt it is on Windows or Linux either.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

N/A

### Is there a reason why this should be core and not an add-on in the asset library?

Affects export templates.

**Comment by Calinou** [comment]
This may be worth implementing on desktop platforms too, as Windows/macOS/Linux all support this in some way.

**Edit:** An example can be seen in the Windows version of [Unleashed Recompiled](https://github.com/hedge-dev/UnleashedRecomp#quality-of-life-options).

**Comment by ztc0611** [comment]
I'm still looking but it actually appears Android might not support it? The API I linked isn't accessible to third party developers and it's the only lead I have as of yet.

Also, any hints about what it might be called on desktop platforms? 

-macOS doesn't use the same signal, only catalyst apps. 
-Windows is so free form about audio that I'm unsure about if it would have any way to distinguish between media sources. (ie: would a system alert sound trip the function if it exists?)
- Don't even have the faintest clue about how anything with desktop Linux works to be honest.

**Comment by ztc0611** [comment]
Done with iOS, working on Android. Unfortunately the way Godot registers for audio on Android makes `AudioManager.isMusicActive()` always return `true`. Not sure what the workaround for that should look like. This behavior is caused by `SL_PLAYSTATE_PLAYING` being set in `platform/android/audio_driver_opensl.cpp`.

**Comment by astillich** [comment]
On Linux you could use D-BUS [MPRIS](https://specifications.freedesktop.org/mpris-spec/latest/Player_Interface.html) Player's Playback_Status property. There doesn't seem to be a corresponding signal so you'd need to poll for changes.

**Comment by ztc0611** [comment]
It appears like it's also going to be that way for Android, so it won't be the first.

**Comment by m4gr3d** [comment]
I question whether this should be part of the core repo since the functionality could instead be provided by a Godot plugin.

@ztc0611 Can you investigate if you can achieve the same level of functionality using a Godot plugin, if so, I'd use that approach instead.

**Comment by ztc0611** [comment]
I disagree. Dealing with background media is an important concern on mobile platforms, I think this shouldn't be a plugin, even if it is possible to be one. iOS has built in Project Settings in the Godot editor to deal with mixing with audio, this is an obvious expansion of that. [secondaryAudioShouldBeSilencedHint](https://developer.apple.com/documentation/avfaudio/avaudiosession/1616600-secondaryaudioshouldbesilencedhi) is listed as something to be potentially used with the `ambient` setting for games. Godot is actually set to `ambient` for iOS [by default](https://docs.godotengine.org/en/stable/classes/class_projectsettings.html#class-projectsettings-property-audio-general-ios-session-category). 

Additionally, on Android if there is no alternative to `AudioManager.isMusicActive()`, there needs to be core-level workaround for this anyways. The issue with `isMusicActive()` getting a false positive due to Godot behavior is outlined in my PR.

**Comment by m4gr3d** [comment]
> Dealing with background media is an important concern on mobile platforms

It's an important concern which is why the options to handle that should be within their own plugin so that plugin can afford to go in depth feature wise.

To use Android as an example, there are [a whole range of options](https://developer.android.com/media/optimize/audio-focus) for how to deal with this concept of audio focus. The apis you're proposing are nowhere enough with giving developers the tools to properly handle it and trying to do so would increase your proposal and PR by an order of complexity, complexity that ends up affecting all Godot developers even if they don't make use of it, and complexity that the Godot team has to support going forward.

On the other hand, using a Godot plugin allows you to make that plugin as simple or as rich feature wise as you want without affecting the complexity of the core project or end users that don't make use of that feature.

Don't see this suggestion as looking down on plugins; there are powerful elements of the Godot ecosystem that should be leveraged where it makes sense (e.g: even a core capability like physics can be replaced via a [godot plugin](https://github.com/godot-jolt/godot-jolt)).

**Comment by ztc0611** [comment]
Then maybe the Android version should be a plugin. This is native IOS functionality with direct and easy functions to access, and important to cut down on jank imo. There are plenty of things in the engine that added far more complexity for one platform. (Global menu bars on macOS come to mind.) 

There is essentially zero chance this will ever be discovered as a plugin, not many people are going to google `Godot mix audio with iOS system audio`, especially since many major apps like Twitter just don't even bother, a lot of people don't even know it's possible for mobile apps to share audio in this way.

**Comment by adenine-dev** [comment]
This would be very nice for the idle game i'm working on so 👍

**Comment by Calinou** [comment]
@adenine-dev Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by jcarnaxide** [comment]
I would like to add to this request, that it would be convenient if there was a way for a game to specify, possibly via project settings, if it would like to shift audio focus to the game.

I find many mobile applications, can shift audio focus to themselves, and sometimes possibly even back to the previous application when minimizing the application.

An example of the behaviour I am looking for:

Spotify is currently playing music
I launch my Godot mobile game
Spotify music is paused, my mobile game audio starts
I minimize my godot mobile game
(Optional) Spotify music continues.

Ideally this can be implemented for both iOS and Android for my use case, but honestly if it only worked for one just to start, I would accept whatever I can get 😄
